Add a select() override that takes no arguments. Allow select(…) to work with generic selectors.

dloehr opened the pull request:
- Add a `select()` override that takes no arguments and returns the entire state
- Improve the way `select(…)` works with generic selectors (where the
  return type depends on the types of the arguments), so that the
  argument types are taken into consideration.

danielnixon commented:
Thanks @dloehr ! Regarding the build failures:
>ERROR: /home/travis/build/agiledigital/typed-redux-saga/src/index.ts:27:11 - interface name must start with a capitalized I

I don't like this tslint rule. I'm happy for you to disable it project-wide in the tslint config if you want. I'll be porting to eslint soon anyway.

> ERROR: /home/travis/build/agiledigital/typed-redux-saga/src/index.ts:27:45 - Interface has only a call signature — use `type Selector<Args extends any[], R> = (state: any, ...args: Args) => R` instead.

This one seems reasonable to be, but I'm happy for you to disable it for this line using a tslint ignore comment if you want.

> ERROR: /home/travis/build/agiledigital/typed-redux-saga/src/index.ts:45:3 - Unexpected if, use a conditional expression (ternary operator) instead.

We can just disable this one with a tslint ignore comment too.

dloehr commented:
I fixed the build errors. I also found that my first attempt didn't work with TypeScript 3.7.0-beta; I've adjusted it so that it works with both 3.6.3 and 3.7.0-beta.
